expr_checker.py

```diff
@@ -0,0 +1,130 @@
+    
+    for arg in kwargs.values():
+        check_type("arguments", arg)
+
+    return check_type("returned", func(*args, **kwargs))
+
+
+def check_ro(obj):
+    if type(obj) in _readonly:
+        return copy(obj)
+    else:
+        return obj
+
+
+class NodeChecker(ast.NodeTransformer):
+    def __init__(self, check_fn, check_type_fn, allowed_attr):
+        self.check_fn = check_fn.__name__
+        self.check_type_fn = check_type_fn.__name__
+        self.allowed_attr = allowed_attr
+
+        self.reserved_name = [self.check_fn, self.check_type_fn, "check_ro"]
+
+        super().__init__()
+
+    def visit_Call(self, node):
+        list(map(self.visit, node.args))
+
+        if isinstance(node.func, ast.Attribute):
+            if node.func.attr not in self.allowed_attr:
+                raise ValueError(
+                    f"safe_eval doesn't allow you to read {node.func.attr}")
+
+            name = copy(node.func.value)
+            node.func.value.__class__ = ast.Call
+            node.func.value.func = ast.Name("check_ro", ctx=ast.Load())
+            node.func.value.args = [name]
+            node.func.value.keywords = []
+
+        else:
+            self.generic_visit(node.func)
+
+        return ast.Call(
+            func=ast.Name(self.check_fn, ctx=ast.Load()),
+            args=[node.func, ast.Name(
+                self.check_type_fn, ctx=ast.Load())] + node.args,
+            keywords=node.keywords
+        )
+
+    def visit_FunctionDef(self, node):
+        if node.name in self.reserved_name:
+            raise NameError(f"safe_eval: {node.name} is a reserved name")
+
+        return node
+
+    def visit_Name(self, node):
+        if node.id in self.reserved_name:
+            raise NameError(f"safe_eval: {node.id} is a reserved name")
+
+        return node
+
+    def visit_Attribute(self, node):
+        if node.attr not in self.allowed_attr:
+            raise ValueError(
+                f"safe_eval doesn't allow you to read {node.attr}")
+
+        if isinstance(node.ctx, ast.Load):
+            return ast.Call(
+                func=ast.Name(self.check_type_fn, ctx=ast.Load()),
+                args=[ast.Constant("attribute"), node],
+                keywords=[]
+            )
+
+        elif isinstance(node.ctx, ast.Store):
+            return ast.Call(
+                func=ast.Name("check_ro", ctx=ast.Load()),
+                args=[node.value],
+                keywords=[]
+            )
+
+        elif isinstance(node.ctx, ast.Del):
+            raise ValueError("safe_eval: You can't delete attribute")
+
+    def visit_Assign(self, node):
+        list(map(self.visit, node.targets))
+
+        return ast.Call(
+            func=ast.Name(self.check_type_fn, ctx=ast.Load()),
+            args=[ast.Constant("assignation"), node.value],
+            keywords=[]
+        )
+
+
+def expr_checker(expr, allowed_attr, readonly, check_type=ast_default_check_type, check_function=ast_default_test):
```

Comment:
I think 2 of the 3 removed comments were about using a function instead of "allowed_attr" and "readonly" (I can not find them anymore on the PR)

so instead of `allowed_attr`:
```python
def __ast_check_getattr(obj, key, value):
    pass
```
instead of `readonly`:
```python
def __ast_check_setattr(obj, key, value):
    pass
```

so we will have more flexibility :)

Reply:
Well yes but it's make it very complicated for me to check the types. I can only return one node per visit. For exemple when I'm setting an attribute I have to check the type (so returning an ast.Call) and I'll also need to call `check_setattr`.

The only idea I have is to drop the `value` argument :thinking:

Reply:
for `check_setattr`, you can call a function that will call `__ast__check_setattr`, then `check_type`, and then `setattr` :) so you have only 1 node with a function call, that will make the other calls (IDK if it's clear)